src/python/pants/backend/cc/dependency_inference/rules.py

```diff
@@ -0,0 +1,177 @@
+            stripped_files_with_multiple_owners[stripped_file.value].update(
+                {stripped_files_to_addresses[stripped_file.value], tgt.address}
+            )
+        else:
+            stripped_files_to_addresses[stripped_file.value] = tgt.address
+
+    # Remove files with ambiguous owners.
+    for ambiguous_stripped_f in stripped_files_with_multiple_owners:
+        stripped_files_to_addresses.pop(ambiguous_stripped_f)
+
+    mapping_not_stripped = {tgt[CCSourceField].file_path: tgt.address for tgt in cc_targets}
+
+    return CCFilesMapping(
+        mapping=FrozenDict(sorted(stripped_files_to_addresses.items())),
+        ambiguous_files=FrozenDict(
+            (k, tuple(sorted(v))) for k, v in sorted(stripped_files_with_multiple_owners.items())
+        ),
+        mapping_not_stripped=FrozenDict(mapping_not_stripped),
+    )
+
+
+@dataclass(frozen=True)
+class CCIncludeDirective:
+    path: str
+    system_paths_only: bool  # True if include used `<foo.h>` instead of `"foo.h"`.
+
+
+def parse_includes(content: str) -> frozenset[CCIncludeDirective]:
+    includes: set[CCIncludeDirective] = set()
+    for line in content.splitlines():
+        m = INCLUDE_REGEX.match(line)
+        if m:
+            if m.group(2):
+                includes.add(CCIncludeDirective(m.group(2)[1:-1], False))
+            elif m.group(3):
+                includes.add(CCIncludeDirective(m.group(3)[1:-1], True))
+    return frozenset(includes)
+
+
+@rule
+async def infer_cc_source_dependencies(
+    request: InferCCDependenciesRequest,
+    cc_files_mapping: CCFilesMapping,
+    cc_infer: CCInferSubsystem,
+) -> InferredDependencies:
+    if not cc_infer.includes:
+        return InferredDependencies([])
+
+    address = request.sources_field.address
+    wrapped_tgt = await Get(WrappedTarget, Address, address)
+    explicitly_provided_deps, hydrated_sources = await MultiGet(
+        Get(ExplicitlyProvidedDependencies, DependenciesRequest(wrapped_tgt.target[Dependencies])),
+        Get(HydratedSources, HydrateSourcesRequest(request.sources_field)),
+    )
+
+    digest_contents = await Get(DigestContents, Digest, hydrated_sources.snapshot.digest)
+    assert len(digest_contents) == 1
+    file_content = digest_contents[0]
+    file_path = PurePath(file_content.path)
+
+    includes = parse_includes(file_content.content.decode())
+
+    result: OrderedSet[Address] = OrderedSet()
+    for include in includes:
+        # Skip system-path includes.
+        if include.system_paths_only:
+            continue
+
+        # First try to resolve the include's path against the same directory where the file is.
+        maybe_relative_file_path = file_path.parent.joinpath(include.path)
+        maybe_relative_address = cc_files_mapping.mapping_not_stripped.get(
+            str(maybe_relative_file_path)
+        )
+        if maybe_relative_address:
+            result.add(maybe_relative_address)
```

Comment:
This should either be included in the ambiguity checks, or shortcircuit after matching probably... otherwise you might get two copies.

Reply:
Probably short circuit since C/C++ compilers do not search the include path if there is a relative path match.